Implemented Storage interface in the class StorageImpl

DanaKhromenko opened the pull request.

DanaKhromenko commented:
Fixed Travis's error: WhitespaceAround: '{' is not preceded with whitespace.

RomanDubovskyi commented on src/main/java/core/basesyntax/impl/StorageImpl.java:
```diff
@@ -1,19 +1,144 @@
+    private K[] keys;
+    private V[] values;
+
+    public StorageImpl() {
+        currentLength = 0;
+    }
```
ініціалізуй keys i values в конструкторі і зможеш прибрати  initializeArraysIfNeeded і перевірки їх на нал в путі і геті

DanaKhromenko replied:
> ініціалізуй keys i values в конструкторі і зможеш прибрати initializeArraysIfNeeded і перевірки їх на нал в путі і геті

Имеешь в ввиду так?
public <K, V> StorageImpl() {
   currentLength = 0;
}

Да, это помешает пользователю класса передавать параметры другого типа, согласна. Но как это поможет мне инициализировать массив? Я не знаю, как обратиться к типам K и V. Как я могу получить их тип на этапе конструктора?
Нельзя ни так:
   keys = (K[]) Array.newInstance(K, MAX_LENGTH);
ни так:
   values = new V[MAX_LENGTH];

RomanDubovskyi commented on src/main/java/core/basesyntax/impl/StorageImpl.java:
```diff
@@ -1,19 +1,144 @@
+        return false;
+    }
+
+    private boolean replacedValueByExistingKey(K key, V value) {
```
ці методи не потрібні, можна просто в путі ітерувати масив поки не дойдемо до поточного розміру - якщо під час цього зустріли однакові ключі ріплейсим value і робимо return з методі в іншому випадку просто сетимо ключ і значення на індекс currentSize

DanaKhromenko replied:
> ці методи не потрібні, можна просто в путі ітерувати масив поки не дойдемо до поточного розміру - якщо під час цього зустріли однакові ключі ріплейсим value і робимо return з методі в іншому випадку просто сетимо ключ і значення на індекс currentSize

Что происходит у меня в методе "put":
1. Если ключ существует: "replacedValueByExistingKey", где я, как ты и сказал, обхожу в цикле и, если нахожу такой ключ, реплейсю, после чего выхожу и из "replacedValueByExistingKey", и из "put".
2. Если ключ не существует: "replacedValueByExistingKey", где я в цикле обхожу, не нахожу ключ, выхожу из "replacedValueByExistingKey". Потом проверяю, позволяет ли размер хранилища добавить туда что-то. Если позволяет, вызываю метод "addNewRowToStorage", где по индексу добавляю ключ и значение.

Итого, я не понимаю, в чём проблема? Ты говоришь, что лучше будет не разбивать "put", "replacedValueByExistingKey" и "addNewRowToStorage", которые несут разный функционал, на три разных метода? Говоришь, что я должна делать всё то же самое, но просто в методе "put"? Это ведь просто сделает его менее удобочитаемым и более громоздким. Разве не так?

RomanDubovskyi commented on src/main/java/core/basesyntax/impl/StorageImpl.java:
```diff
@@ -1,19 +1,144 @@
+            return values == null ? null : values[0];
+        }
+        for (int i = 0; i < currentLength; i++) {
+            if ((keys[i] == null) == (key == null)
```
```suggestion
            if ((keys[i] == key)
```
якщо два об'єкта нал - порівняння по посиланню відпрацюють коректно, якщо ні у насє друга частина іфу

DanaKhromenko replied:
Если сделать как ты указал, получаю ошибку тестов.

У меня ведь целиком условие так выглядит:
if ((keys[i] == null) == (key == null)
                    && (key == null || keys[i].equals(key))) {

И если в первой половине я поставлю "(keys[i] == key)", то это значит, что ключи должны совпадать по ссылке, а тогда валится тест "getElementWhenKeyIsObject", где сравнивают "elementTwo" и storage.get(sameSecondCat)".

Если же твоё "if ((keys[i] == key)" заменить на "if ((keys[i].equals(key))", то валится тест "getElementWithKeyNull", так как при key = null срабатывает ошибка "java.lang.NullPointerException: Cannot invoke "Object.equals(Object)" because "this.keys[i]" is null"

DanaKhromenko commented:
- replaced method "addedValueToNullArray()" because after changing the initialization, there is no more need in it;
- erased it's callings in the code;
- removed methods "replacedValueByExistingKey" and "addNewRowToStorage", replacing their functions into the "put" method;
- replaced statement "(keys[i] == null) == (key == null) && (key == null || keys[i].equals(key))" with the statement "keys[i] == key || (keys[i] != null && keys[i].equals(key))".
